File: src/main/java/com/jcloisterzone/game/capability/GamblersLuckCapability.java
package com.jcloisterzone.game.capability;

import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.feature.City;
import com.jcloisterzone.feature.Feature;
import com.jcloisterzone.feature.Field;
import com.jcloisterzone.feature.GamblersLuckShield;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.capability.FamiliesCapability;
import com.jcloisterzone.game.Token;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTile;
import com.jcloisterzone.random.RandomGenerator;

import static com.jcloisterzone.XMLUtils.attributeStringValue;

import io.vavr.Tuple2;
import io.vavr.collection.HashMap;
import io.vavr.collection.Map;
import io.vavr.collection.Stream;

import org.w3c.dom.Element;

/**
 * Capability model is {@code Map<FeaturePointer, String>} - Gambler's Luck Shield Tokens,
 *   key is Gambler's Luck Shield Token id or null if nothing is placed
 */
public final class GamblersLuckCapability extends Capability<Map<FeaturePointer, Tuple2<GamblersLuckCapability.GamblersLuckShieldToken,Integer>>> {
	
	public enum GamblersLuckShieldToken implements Token {
		GAMBLERSLUCKSHIELD_0,
		GAMBLERSLUCKSHIELD_1,
		GAMBLERSLUCKSHIELD_2,
		GAMBLERSLUCKSHIELD_3,
		GAMBLERSLUCKSHIELD_X
	}
	
	private static final long serialVersionUID = 1L;
	
	@Override
    public Feature initFeature(GameState state, String tileId, Feature feature, Element xml) {
        if (feature instanceof City) {
    	    if (!("".equals(attributeStringValue(xml, "gamblers-luck-shield", "")))) {
                String family = "unknown";
                if (xml.hasAttribute("family")) {
                    family = xml.getAttribute("family");
                }
                feature = ((City) feature).putModifier(FamiliesCapability.FAMILY, family);
    	    }
        }
        return feature;
    }

	@Override
	public GameState onStartGame(GameState state, RandomGenerator random) {
		return setModel(state, HashMap.empty());
	}
	
	public Stream<Tuple2<FeaturePointer,GamblersLuckShield>> getPlacedTileGamblersLuckShields(GameState state) {
		return getGamblersLuckShieldsOnPlacedTile(state, state.getLastPlaced());
	}

	public Stream<Tuple2<FeaturePointer,GamblersLuckShield>> getGamblersLuckShieldsOnPlacedTile(GameState state, PlacedTile pt) {
		return state.getTileFeatures2(pt.getPosition(), GamblersLuckShield.class);
	}

	public boolean hasPlacedTileGamblersLuckShields(GameState state) {
		return getPlacedTileGamblersLuckShields(state).length()!=0;
	}
}

File: src/main/ts/com/jcloisterzone/game/capability/GamblersLuckCapability.ts
import { HashMap, type Map as VMap } from "../../../../io/vavr/Map.js";
import type { Stream } from "../../../../io/vavr/SeqTypes.js";
import { Tuple2 } from "../../../../io/vavr/Tuple.js";
import { JavaEnum } from "../../../../lang/JavaEnum.js";
import type { FeaturePointer } from "../../board/pointer/FeaturePointer.js";
import { GamblersLuckShield } from "../../feature/GamblersLuckShield.js";
import type { RandomGenerator } from "../../random/RandomGenerator.js";
import { Capability } from "../Capability.js";
import type { Token } from "../Token.js";
import type { GameState } from "../state/GameState.js";
import type { PlacedTile } from "../state/PlacedTile.js";

type Model = VMap<FeaturePointer, Tuple2<GamblersLuckCapability.GamblersLuckShieldToken, number>>;

/** Gamblers' Luck promo. Model: placed shield tokens by feature pointer. */
export class GamblersLuckCapability extends Capability<Model> {
  override onStartGame(state: GameState, random: RandomGenerator): GameState {
    return this.setModel(state, HashMap.empty());
  }

  getPlacedTileGamblersLuckShields(
    state: GameState,
  ): Stream<Tuple2<FeaturePointer, GamblersLuckShield>> {
    return this.getGamblersLuckShieldsOnPlacedTile(state, state.getLastPlaced()!);
  }

  getGamblersLuckShieldsOnPlacedTile(
    state: GameState,
    pt: PlacedTile,
  ): Stream<Tuple2<FeaturePointer, GamblersLuckShield>> {
    return state.getTileFeatures2(pt.getPosition(), GamblersLuckShield);
  }

  hasPlacedTileGamblersLuckShields(state: GameState): boolean {
    return this.getPlacedTileGamblersLuckShields(state).length() !== 0;
  }
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace GamblersLuckCapability {
  export class GamblersLuckShieldToken extends JavaEnum implements Token {
    static readonly GAMBLERSLUCKSHIELD_0 = new GamblersLuckShieldToken("GAMBLERSLUCKSHIELD_0", 0);
    static readonly GAMBLERSLUCKSHIELD_1 = new GamblersLuckShieldToken("GAMBLERSLUCKSHIELD_1", 1);
    static readonly GAMBLERSLUCKSHIELD_2 = new GamblersLuckShieldToken("GAMBLERSLUCKSHIELD_2", 2);
    static readonly GAMBLERSLUCKSHIELD_3 = new GamblersLuckShieldToken("GAMBLERSLUCKSHIELD_3", 3);
    static readonly GAMBLERSLUCKSHIELD_X = new GamblersLuckShieldToken("GAMBLERSLUCKSHIELD_X", 4);
    private static readonly VALUES: readonly GamblersLuckShieldToken[] = [
      GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_0,
      GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_1,
      GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_2,
      GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_3,
      GamblersLuckShieldToken.GAMBLERSLUCKSHIELD_X,
    ];
    static values(): readonly GamblersLuckShieldToken[] {
      return GamblersLuckShieldToken.VALUES;
    }
  }
}

Capability.register(GamblersLuckCapability);
